src/hardware/serialport/serialmouse.cpp

```diff
@@ -0,0 +1,493 @@
+		// Byte 2:  Y7 Y6 Y5 Y4 Y3 Y2 Y1 Y0
+
+		uint8_t bt = has_3rd_button ? ((~buttons) & 7) : ((~buttons) & 3);
+
+		packet[0]  = static_cast<uint8_t>(0x80 | ((bt & 1) << 2) |
+                                                 ((bt & 2) >> 1) | ((bt & 4) >> 1));
+		packet[1]  = ClampDelta(delta_x);
+		packet[2]  = ClampDelta(-delta_y);
+		packet_len = 3;
+		xmit_2part = true; // next part contains mouse movement since
+		                   // the start of the 1st part
+
+	} else {
+		LogUnimplemented();
+	}
+
+	ClearCounters();
+
+	// send packet
+	xmit_idx       = 0;
+	another_button = false;
+	another_move   = false;
+	SetEventRX();
+}
+
+void CSerialMouse::StartPacketPart2()
+{
+	// port settings are valid at this point
+
+	if (type == MouseType::MouseSystems) {
+		//          -- -- -- -- -- -- -- --
+		// Byte 3:  X7 X6 X5 X4 X3 X2 X1 X0
+		// Byte 4:  Y7 Y6 Y5 Y4 Y3 Y2 Y1 Y0
+
+		packet[0] = ClampDelta(delta_x);
+		packet[1] = ClampDelta(-delta_y);
+
+		packet_len = 2;
+		xmit_2part = false;
+	} else {
+		LogUnimplemented();
+	}
+
+	ClearCounters();
+
+	// send packet
+	xmit_idx     = 0;
+	another_move = false;
+	SetEventRX();
+}
+
+void CSerialMouse::SetEventTX()
+{
+	setEvent(SERIAL_TX_EVENT, bytetime / smooth_div);
+}
+
+void CSerialMouse::SetEventRX()
+{
+	setEvent(SERIAL_RX_EVENT, bytetime / smooth_div);
+}
+
+void CSerialMouse::SetEventTHR()
+{
+	setEvent(SERIAL_THR_EVENT, bytetime / 10);
+}
+
+void CSerialMouse::LogUnimplemented() const
+{
+	LOG_ERR("MOUSE (COM%d): Missing implementation", port_num);
+}
+
+uint8_t CSerialMouse::ClampDelta(const int32_t delta) const
+{
+	return static_cast<uint8_t>(std::clamp(delta,
+	                                       static_cast<int32_t>(INT8_MIN),
+	                                       static_cast<int32_t>(INT8_MAX)));
+}
+
+void CSerialMouse::handleUpperEvent(const uint16_t event_type)
+{
+	if (event_type == SERIAL_TX_EVENT) {
+		ByteTransmitted(); // tx timeout
+	} else if (event_type == SERIAL_THR_EVENT) {
+		ByteTransmitting();
+		SetEventTX();
+	} else if (event_type == SERIAL_RX_EVENT) {
+		// check for bytes to be sent to port
+		if (CSerial::CanReceiveByte()) {
+			if (send_ack) {
+				send_ack = false;
+				StartPacketId();
+			} else if (xmit_idx < packet_len) {
+				CSerial::receiveByte(packet[xmit_idx++]);
+				if (xmit_idx >= packet_len && xmit_2part)
+					StartPacketPart2();
+				else if (xmit_idx >= packet_len &&
+				         (another_move || another_button))
+					StartPacketData();
+				else
+					SetEventRX();
+			}
+		} else
+			SetEventRX();
+	}
+}
+
+void CSerialMouse::updatePortConfig(const uint16_t divider, const uint8_t lcr)
+{
+	AbortPacket();
+
+	// Check whether port settings match mouse protocol,
+	// to prevent false device detections by guest software
+
+	port_valid = true;
+
+	const uint8_t port_byte_len = static_cast<uint8_t>((lcr & 0x3) + 5);
+	const bool one_stop_bit     = !(lcr & 0x4);
+	const uint8_t parity_id     = static_cast<uint8_t>((lcr & 0x38) >> 3);
```

Comment:
`const auto ..` for all three.

(Might as well let the right-hand-size type-specifier be used, and make the code a bit DRYer)

Reply:
Done.